CoreScriptsRoot/Modules/VR/Panel3D.lua

```diff
@@ -0,0 +1,760 @@
+	if UserInputService.OverrideMouseIconBehavior == Enum.OverrideMouseIconBehavior.ForceShow then
+		return true
+	end
+	--if ForceHide, the cursor is definitely NOT visible
+	if UserInputService.OverrideMouseIconBehavior == Enum.OverrideMouseIconBehavior.ForceHide then
+		return false
+	end
+	--Otherwise, we need to check if the developer set MouseIconEnabled=false
+	if UserInputService.MouseIconEnabled and UserInputService.OverrideMouseIconBehavior == Enum.OverrideMouseIconBehavior.None then
+		return true
+	end
+	return false
+end
+
+--End of cursor autohiding methods
+
+
+--Panel class implementation
+local Panel = {}
+Panel.mt = { __index = Panel }
+function Panel.new(name)
+	local instance = {
+		name = name,
+
+		part = nil,
+		gui = nil,
+
+		width = 1,
+		height = 1,
+
+		isVisible = false,
+		isEnabled = false,
+		panelType = Panel3D.Type.None,
+		pixelScale = 1,
+		showCursor = true,
+		shouldFindLookAtGuiElement = false,
+
+		linkedTo = nil,
+		subpanels = {},
+
+		transparency = 1,
+		forceShowUntilLookedAt = false,
+		isLookedAt = false,
+		isOffscreen = true,
+		lookAtPixel = Vector2.new(-1, -1),
+		lookAtDistance = math.huge,
+		lookAtGuiElement = nil,
+		isClosest = true
+	}
+
+	if panels[name] then
+		error("A panel by the name of " .. name .. " already exists.")
+	end
+	panels[name] = instance
+
+	return setmetatable(instance, Panel.mt)
+end
+
+--Panel accessor methods
+function Panel:GetPart()
+	if not self.part then
+		self.part = Util.Create "Part" {
+			Name = self.name,
+			Parent = nil,
+
+			Transparency = 1,
+
+			CanCollide = false,
+			Anchored = true,
+			Archivable = false,
+
+			Size = Vector3.new(1, 1, 1)
+		}
+	end
+	return self.part
+end
+
+function Panel:GetGUI()
+	if not self.gui then
+		local part = self:GetPart()
+		self.gui = Util.Create "SurfaceGui" {
+			Parent = CoreGui,
+			Adornee = part,
+			Active = true,
+			ToolPunchThroughDistance = 1000,
+			CanvasSize = Vector2.new(0, 0),
+			Enabled = self.isEnabled
+		}
+		--todo: remove this pcall when api is live
+		pcall(function()
+			self.gui.AlwaysOnTop = true
+		end)
+	end
+	return self.gui
+end
+
+function Panel:FindHoveredGuiElement(elements)
+	local x, y = self.lookAtPixel.X, self.lookAtPixel.Y
+	for i, v in pairs(elements) do
+		local minPt = v.AbsolutePosition
+		local maxPt = v.AbsolutePosition + v.AbsoluteSize
+		if minPt.X <= x and maxPt.X >= x and
+		   minPt.Y <= y and maxPt.Y >= y then
+			return v, i
+		end
+	end
+end
+--End of panel accessor methods
+
+
+--Panel update methods
+function Panel:SetPartCFrame(cframe)
+	if not self.part then
+		return
+	end
+	self.part.CFrame = cframe * CFrame.new(0, 0, -0.5)
+end
+
+function Panel:SetEnabled(enabled)
+	if self.isEnabled == enabled then
+		return
+	end
+
+	self.isEnabled = enabled
+	if not enabled and (not self.part or not self.gui) then
+		return
+	else
+		self:GetPart()
+		self:GetGUI()
+	end
+	if enabled then
+		self.part.Parent = workspace.CurrentCamera --todo: Perhaps this can change soon.
```

Comment:
Oh hey. I think my AlwaysOnTop fixes in the C++ side should fix this up. I'll experiment with removing this.